mmv1/third_party/cai2hcl/common/uber_converter.go

```diff
@@ -0,0 +1,46 @@
+package common
+
+import (
+	"github.com/GoogleCloudPlatform/terraform-google-conversion/v5/caiasset"
+)
+
+// Converter which aggregates all service-specific converters in the same interface.
+type UberConverter struct {
```

Comment:
I would be interested in potentially calling this `Converter` and having a `NewConverter` function, so that the implementation more closely matches tfplan2cai: https://github.com/GoogleCloudPlatform/terraform-google-conversion/blob/main/tfplan2cai/converters/google/convert.go

That said, I'm also open to alternatives - mostly I think it would be good to converge on a more similar implementation between the two halves over time so that we can reduce cognitive load moving between them.

What would you think?

Reply:
In Cai2hcl Converter is an [interface](https://github.com/GoogleCloudPlatform/terraform-google-conversion/blob/main/cai2hcl/common/converter.go), so we cannot call it so.

UberConverter is on one hand implements this Converter interface, but on the other - itself a new interface for group of converters... The naming is frustrating, I agree.

Reply:
Yeah, the interface would need to be renamed as well - or we would need to stop using interfaces like this. Overall I'm concerned that there's a lot of complexity / nesting going on that ends up obfuscating what's happening, and I'm not sure it's benefitting us. But I could totally be forgetting or not considering something!

FWIW my inclination would be that we standardize towards how tfplan2cai works (having a [ResourceConverter struct](https://github.com/GoogleCloudPlatform/terraform-google-conversion/blob/ad14a1096e314d4245d5467fe04ab302bea97630/tfplan2cai/converters/google/resources/services/compute/compute_forwarding_rule.go#L50) with a predictable structure - this is similar to how [Terraform resources are implemented](https://github.com/hashicorp/terraform-provider-google/blob/20615b0badd5121b90878658b185c62a6dd64416/google/services/compute/resource_compute_forwarding_rule.go#L53).

This would basically mean replacing the New*Converter functions with a `ResourceConverter` that returns a standard type which knows the Terraform type it converts to.

We could than have a single map that of all asset types to relevant resource converters, similar to [resource_converters.go](https://github.com/GoogleCloudPlatform/terraform-google-conversion/blob/0ff3c33174570270e0bc937ebadae0b6ee80d6ed/tfplan2cai/converters/google/resources/resource_converters.go#L72) in tfplan2cai and [provider_mmv1_resources.go](https://github.com/hashicorp/terraform-provider-google/blob/20615b0badd5121b90878658b185c62a6dd64416/google/provider/provider_mmv1_resources.go#L123) in the provider. It's not elegant but it's very clear what's going on, and we would be able to avoid the indirection of the ConverterMap (and I think the names of things could be clearer as well.)

I have a vague memory that there was a reason we didn't do this previously - possibly a circular dependency? - but I don't remember why it was a problem for cai2hcl and not for tfplan2cai. And that might have been related to a completely different decision.

As before, I'm open to alternatives (and even hypothetically keeping the current structure) if there's a clear benefit or a blocker of some kind that outweighs standardizing on our other libraries.